components/eam/src/chemistry/utils/tracer_data.F90

```diff
@@ -1602,6 +1650,183 @@ subroutine read_za_trc( fid, vid, loc_arr, strt, cnt, file, order )
+    use phys_grid,        only : pcols, begchunk, endchunk, get_ncols_p, get_rlat_all_p, get_rlon_all_p 
+    use mo_constants,     only : pi
+    use dycore,           only : dycore_is              
+    use polar_avg,        only : polar_average
+
+    implicit none
+    type(file_desc_t), intent(in) :: fid
+    type(var_desc_t),  intent(in) :: vid
+    integer,           intent(in) :: strt(:), cnt(:)
+    integer,           intent(in) :: order(2)
+    real(r8),          intent(out):: loc_arr(:,:,:)
+    type (trfile),     intent(in) :: file
+    !!
+    type(var_desc_t),  intent(in), optional :: vid_srf
+    integer :: cnt_srf(2)
+    integer :: strt_srf(2)
+    !!
+    type(interp_type) :: lat_wgts
+    real(r8) :: to_lats(pcols), to_lons(pcols), wrk(pcols)
+    real(r8), allocatable, target :: wrk2d(:,:)
+    real(r8), allocatable, target :: wrksrf(:)
+    real(r8), pointer :: wrk2d_in(:,:)
+    integer :: c, k, ierr, ncols
+
+     nullify(wrk2d_in)
+     allocate( wrk2d(cnt(1),cnt(2)), stat=ierr )
+     if( ierr /= 0 ) then
+        write(iulog,*) 'read_2d_trc: wrk2d allocation error = ',ierr
+        call endrun
+     end if
+
+     if(order(1)/=1 .or. order(2)/=2 .or. cnt(1)/=file%nlat .or. cnt(2)/=file%nlev) then
+        allocate( wrk2d_in(file%nlat, file%nlev), stat=ierr )
+        if( ierr /= 0 ) then
+           write(iulog,*) 'read_2d_trc: wrk2d_in allocation error = ',ierr
+           call endrun
+        end if
+     end if
+    !!
+    ierr = pio_get_var( fid, vid, strt, cnt, wrk2d )
+    !!
+    if (file%linoz_v3) then
+            !!since io reads in global data for every thread
+            !!and interpolate to local chunk, we can do surface,
+            !!polar, and surface padding preprocessfor every variable 
+            !!before interpolation
+            !!it is put here rather than the data formation process
+            !!to prevent forget when producing forcing data
+                !read in order of 0.1hPa~985hPa, pad top with 2nd layer
+                wrk2d(:,1)=wrk2d(:,2)
+                !both N/S poles need to pad by nearest data
+                wrk2d(1,:)=wrk2d(2,:)
+                wrk2d(file%nlat,:)=wrk2d(file%nlat-1,:)
+                !!read in srf data to pad surface
+                !!they are inputs from CMIP forcing and other references
+                cnt_srf(1)=cnt(1)
+                cnt_srf(2)=cnt(3)
+                strt_srf(1)=strt(1)
+                strt_srf(2)=strt(3)
+                !!check if vid_srf is present to determine clim variables
+                if (present(vid_srf)) then
+                !!padding for clim terms
+                allocate(wrksrf(cnt(1)), stat=ierr )
+                ierr = pio_get_var( fid, vid_srf, strt_srf, cnt_srf, wrksrf )
+                !!surface padding
+                wrk2d(:,file%nlev)=wrksrf
+                deallocate(wrksrf)
+                else
+                !!padding for derivative terms
+                wrksrf=0.0_r8
```

Comment:
@jinboxie , look like there is no need of this (padding for derivative).
wrksrf in the else block is never allocated. While runs with intel compiler didn't complain, all gnu tests failed due to invalid memory reference.

Reply:
Verified: Commenting out lines 1723-1725 allows gnu tests to run, while having no impact on intel tests.

Reply:
Ok, thank you guys!